memberlist: reresolve members with every 100 nodes on Join

dimitarvdimitrov opened the pull request:
### Description 

This change solves three problems:
1. Context cancelations aren't respected in Join. Sometimes join might take as long as 25 minutes and there is no way to cancel it. Now we check the context on every 100 joined nodes
2. `Join` will attempt to join member IPs that become obsolete between the time discoverMembers runs and memberlist initiates a push-pull with the node. Reresolve `JoinMembers` addresses after joining every 100 nodes. This helps to clean up obsolete IPs that will be tried at the end of the Join procedure
3. (minor) fast joining on startup doesn't respect context cancelations

The side effect of this change is that there will be more DNS resolutions for the JoinMembers.

**Note to reviewers:** `TestJoinMembersWithRetryBackoff` seems to sometimes fail locally, but it should be an unrelated failure. I believe it should be fixed by this https://github.com/grafana/dskit/pull/412.

**Checklist**
- [ ] Tests updated
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

pracucci commented on kv/memberlist/memberlist_client.go:
```diff
@@ -568,41 +568,88 @@ func (m *KV) joinMembersOnStartup(ctx context.Context) bool {
+		successfullyJoined = 0
+	)
+
+	for ; boff.Ongoing(); boff.Wait() {
```
[nit] To keep the code easier to read, I would move the whole content of the `for` loop to the function `joinMembersInBatches()` and rename this function to `joinMembersInBatchesWithRetries()`. In `joinMembersInBatchers()` instead of checking `boff.Ongoing()` you could just check the `ctx.Err()`.

pracucci commented on kv/memberlist/memberlist_client.go:
```diff
@@ -568,41 +568,88 @@ func (m *KV) joinMembersOnStartup(ctx context.Context) bool {
+			joinedInBatch  int
+			err            error
+		)
+		for batchHasMore := true; boff.Ongoing() && batchHasMore; successfullyJoined += joinedInBatch {
```
[nit] To make code easier to read, I move move `successfullyJoined += joinedInBatch` to the end of the for loop and define `joinedInBatch` inside the for loop itself. Keep it outside looks a bit an anti-pattern to me.

pracucci commented on kv/memberlist/memberlist_client.go:
```diff
@@ -568,41 +568,88 @@ func (m *KV) joinMembersOnStartup(ctx context.Context) bool {
-		// We rejoin all nodes, including those that were joined during "fast-join".
-		// This is harmless and simpler.
-		nodes := m.discoverMembers(ctx, m.cfg.JoinMembers)
+func (m *KV) joinMembersBatch(nodes []string, batchSize int, attemptedNodes map[string]bool) (moreLeft bool, successfullyJoined int, lastErr error) {
```
[nit] From a design perspective, passing all `nodes` and `attemptedNodes` looks a bit an anti-pattern to me. I don't think it's responsibility of the function joining a batch of nodes. The caller could build the list of nodes to join, and `joinMembersBatch()` just does it. Also I would take `ctx` in input here, so that you can break the loop as soon as context is cancelled, without waiting for all 100 nodes in the batch to be tried.

pracucci reviewed (approved):
I'm good with the changes, but I think the implementation is a bit more complicated (see it as "harder to follow") than required. A slightly better design may simplify it.

dimitarvdimitrov commented:
thank you for the review @pracucci. I think I addressed your comments. I decided to go with 3 functions: 1) manage retries, 2) manage batches, 3) join a single batch. The code become a bit longer, but I think it's more clear

pstibrany commented on kv/memberlist/memberlist_client.go:
```diff
@@ -568,41 +568,125 @@ func (m *KV) joinMembersOnStartup(ctx context.Context) bool {
+			if attemptedNodes[n] {
+				continue
+			}
+			if len(batch) >= batchSize {
```
nit: should we move this `if` after `batch = append(batch, n)`?

dimitarvdimitrov replied:
then we don't know whether there are `moreAvailableNodes`. If we have filled the batch size and there is another node which wasn't attempted already, then we know. This is covering the edge case where the batch fills with the last item in `nodes`

pstibrany commented on kv/memberlist/memberlist_client.go:
```diff
@@ -568,41 +568,125 @@ func (m *KV) joinMembersOnStartup(ctx context.Context) bool {
+		// Record joined nodes
+		successfullyJoined += joinedInBatch
+		for _, n := range batch {
+			attemptedNodes[n] = true
```
nit: we can move this to the previous `for` loop, when we add node to batch.

dimitarvdimitrov replied:
ah, good point

pstibrany commented on kv/memberlist/memberlist_client.go:
```diff
@@ -568,41 +568,125 @@ func (m *KV) joinMembersOnStartup(ctx context.Context) bool {
+		if ctx.Err() != nil {
+			return successfullyJoined, fmt.Errorf("joining batch: %w", context.Cause(ctx))
+		}
+		// Attempt to join a single node. Complexity shouldn't be different from passing all the node IPs to Join.
```
> Complexity shouldn't be different from passing all the node IPs to Join.

What is this comment trying to say?

dimitarvdimitrov replied:
that it doesn't matter whether we call Join with all nodes once or call it with one node N times. I reworded the comment

pstibrany commented on kv/memberlist/memberlist_client.go:
```diff
@@ -568,41 +568,125 @@ func (m *KV) joinMembersOnStartup(ctx context.Context) bool {
+		level.Warn(logger).Log("msg", "joining memberlist cluster", "attempts", boff.NumRetries()+1, "max_attempts", numAttempts, "err", err)
+	}
+	if err == nil && ctx.Err() != nil {
+		err = fmt.Errorf("joining memberlist: %w", context.Cause(ctx))
```
Does setting `err` here possibly violate the comment saying "joinMembersWithRetries returns an error iff it couldn't successfully join any node."? (What if context is cancelled after we already joined some nodes?)

dimitarvdimitrov replied:
that's right. I updated the comment

pstibrany commented on kv/memberlist/memberlist_client.go:
```diff
@@ -568,41 +568,125 @@ func (m *KV) joinMembersOnStartup(ctx context.Context) bool {
+			MaxRetries: numAttempts,
+		}
+		boff               = backoff.New(ctx, cfg)
+		err                = error(nil)
```
nit:
```suggestion
		err                error
```

pstibrany commented on kv/memberlist/memberlist_client.go:
```diff
@@ -568,41 +568,125 @@ func (m *KV) joinMembersOnStartup(ctx context.Context) bool {
+		}
+		level.Warn(logger).Log("msg", "joining memberlist cluster", "attempts", boff.NumRetries()+1, "max_attempts", numAttempts, "err", err)
+	}
+	if err == nil && ctx.Err() != nil {
```
instead of `ctx.Err`, I would expect that we use `boff.Err()`.

pstibrany commented on kv/memberlist/memberlist_client.go:
```diff
@@ -568,41 +568,125 @@ func (m *KV) joinMembersOnStartup(ctx context.Context) bool {
-	level.Error(m.logger).Log("msg", "joining memberlist cluster failed", "last_error", lastErr, "elapsed_time", time.Since(startTime))
-	return false
+// joinMembersBatch returns an error only if it couldn't successfully join any nodes or if ctx is cancelled.
+func (m *KV) joinMembersBatch(ctx context.Context, nodes []string) (successfullyJoined int, lastErr error) {
```
suggestion: joinMembersBatch -> joinAllMembers

dimitarvdimitrov replied:
i prefer joinMembersBatch because it complements joinMembersInBatches and it's more clear that InBatches uses Batch

pstibrany reviewed (approved):
lgtm overall, comments are not blocking.

dimitarvdimitrov commented:
thanks for the reviews @pracucci, @pstibrany. I will merge this after CI passes

dimitarvdimitrov commented:
there is a race condition with some printing of atomics in the test. I don't expect this was introduced by this PR. I will investigate it and push a fix.

dimitarvdimitrov commented:
i pushed a commit which gets rid of the data race. The race was there when a service was being described. The race detector treats printing a mutex field and using that mutex as a race condition. This can be circumvented by giving the memberlist kv a name, so it doesn't have to be printed. The change is in https://github.com/grafana/dskit/pull/411/commits/ab6e2d1664431cf2144f7d67da3700f565efc7b3
